server/leaderboard/calculate.js

```diff
@@ -1,70 +1,133 @@
+
+  for (let i = 0; i < allChallenges.length; i++) {
+    const challenge = allChallenges[i]
+    if (!solveAmount.has(challenge.id)) {
+      // There are currently no solves
+      challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, 0))
+    } else {
+      challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, solveAmount.get(challenge.id)))
+    }
   }
-}
 
-for (let i = 0; i < allChallenges.length; i++) {
-  const challenge = allChallenges[i]
-  if (!solveAmount.has(challenge.id)) {
-    // There are currently no solves
-    challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, 0))
-  } else {
-    challengeValues.set(challenge.id, utilScores.getScore('dynamic', challenge.points.min, challenge.points.max, solveAmount.get(challenge.id)))
+  for (let i = 0; i < users.length; i++) {
+    const user = users[i]
+    let currScore = 0
+    const solvedChalls = userSolves.get(user.id)
+    if (solvedChalls === undefined) continue // If the user has not solved any challenges, do not add to leaderboard
+    for (let j = 0; j < solvedChalls.length; j++) {
+      // Add the score for the specific solve loaded from the challengeValues array using ids
+      currScore += challengeValues.get(solvedChalls[j])
+    }
+    userScores.push([user.id, user.name, parseInt(user.division), currScore, userLastSolves.get(user.id)])
   }
-}
 
-for (let i = 0; i < users.length; i++) {
-  const user = users[i]
-  let currScore = 0
-  const solvedChalls = userSolves.get(user.id)
-  if (solvedChalls === undefined) continue // If the user has not solved any challenges, do not add to leaderboard
-  for (let j = 0; j < solvedChalls.length; j++) {
-    // Add the score for the specific solve loaded from the challengeValues array using ids
-    currScore += challengeValues.get(solvedChalls[j])
+  return {
+    challengeValues,
+    userScores
   }
-  userScores.push([user.id, user.name, parseInt(user.division), currScore, userLastSolves.get(user.id)])
 }
 
-const sortedUsers = userScores.sort((a, b) => {
+const userCompare = (a, b) => {
   // sort the users by score
   // if two user's scores are the same, sort by last solve time
   const scoreCompare = b[3] - a[3]
   if (scoreCompare !== 0) {
     return scoreCompare
   }
   return a[4] - b[4]
-}).map((user) => user.slice(0, 4))
+}
+
+if (graph) {
+  const samples = calcSamples({ start, end })
+  const output = []
+
+  samples.forEach((sample) => {
+    const { userScores } = calculateScores(sample)
+    output.push({
+      sample,
+      scores: userScores.map((score) => [score[0], score[3]])
+    })
+  })
+
+  parentPort.postMessage({
+    leaderboards: output
+  })
+} else {
+  const nextSample = getNextSample()
+  let sample = Date.now()
+  let isSample = false
+  if ((nextSample - Date.now()) < config.leaderboardUpdateInterval) {
```

Comment:
We should decide whether or not to calculate a new sample based on the time of the previous sample instead of trying to guess based on the current time and when the next sample "should" be.